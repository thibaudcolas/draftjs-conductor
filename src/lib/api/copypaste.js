// @flow
import type { ElementRef } from "react";
import {
  Editor,
  EditorState,
  Modifier,
  convertToRaw,
  convertFromRaw,
  ContentState,
} from "draft-js";
import getFragmentFromSelection from "draft-js/lib/getFragmentFromSelection";

// Custom attribute to store Draft.js content in the HTML clipboard.
const FRAGMENT_ATTR = "data-draftjs-conductor-fragment";

/**
 * Overrides the default copy/cut behavior, adding the serialised Draft.js content to the clipboard data.
 * See also https://github.com/basecamp/trix/blob/62145978f352b8d971cf009882ba06ca91a16292/src/trix/controllers/input_controller.coffee#L415-L422
 * We serialise the editor content within HTML, not as a separate mime type, because Draft.js only allows access
 * to HTML in its paste event handler.
 */
const draftEditorCopyListener = (ref: ElementRef<Editor>, e: Object) => {
  const selection = window.getSelection();
  // Get clipboard content like Draft.js would.
  // https://github.com/facebook/draft-js/blob/37989027063ccc8279bfdc99a813b857549512a6/src/component/handlers/edit/editOnCopy.js#L34
  const fragment = getFragmentFromSelection(ref._latestEditorState);

  // Override the default behavior if there is a selection, and content, and clipboardData is supported (IE11 is out).
  if (selection.rangeCount > 0 && fragment && e.clipboardData) {
    const content = ContentState.createFromBlockArray(fragment.toArray());
    const serialisedContent = JSON.stringify(convertToRaw(content));

    // Create a temporary element to store the selectionâ€™s HTML.
    // See also Rangy's implementation: https://github.com/timdown/rangy/blob/1e55169d2e4d1d9458c2a87119addf47a8265276/src/core/domrange.js#L515-L520.
    const fragmentElt = document.createElement("div");
    // Modern browsers only support a single range.
    fragmentElt.appendChild(selection.getRangeAt(0).cloneContents());
    fragmentElt.setAttribute(FRAGMENT_ATTR, serialisedContent);
    // We set the style property to replicate the browser's behavior of inline styles in rich text copy-paste.
    // In Draft.js, this is important for line breaks to be interpreted correctly when pasted into another word processor.
    // See https://github.com/facebook/draft-js/blob/a1f4593d8fa949954053e5d5840d33ce1d1082c6/src/component/base/DraftEditor.react.js#L328.
    fragmentElt.setAttribute("style", "white-space: pre-wrap;");

    e.clipboardData.setData("text/plain", selection.toString());
    e.clipboardData.setData("text/html", fragmentElt.outerHTML);

    e.preventDefault();
  }
};

export const registerCopySource = (ref: ElementRef<Editor>) => {
  const editorElt = ref.editor;
  const onCopy = draftEditorCopyListener.bind(null, ref);

  editorElt.addEventListener("copy", onCopy);
  editorElt.addEventListener("cut", onCopy);

  return {
    unregister() {
      editorElt.removeEventListener("copy", onCopy);
      editorElt.removeEventListener("cut", onCopy);
    },
  };
};

/**
 * Handles pastes coming from Draft.js editors set up to serialise
 * their Draft.js content within the HTML.
 * This SHOULD NOT be used for stripPastedStyles editor.
 */
export const handleDraftEditorPastedText = (
  html: ?string,
  editorState: EditorState,
) => {
  // Plain-text pastes are better handled by Draft.js.
  // flowlint sketchy-null-string:off
  if (!html) {
    return false;
  }

  const doc = new DOMParser().parseFromString(html, "text/html");
  const fragmentElt = doc.querySelector(`[${FRAGMENT_ATTR}]`);

  // Handle the paste if it comes from draftjs-conductor.
  if (fragmentElt) {
    const fragmentAttr = fragmentElt.getAttribute(FRAGMENT_ATTR);
    let rawContent;

    try {
      // If JSON parsing fails, leave paste handling to Draft.js.
      // There is no reason for this to happen, unless the clipboard was altered somehow.
      // $FlowFixMe
      rawContent = JSON.parse(fragmentAttr);
    } catch (error) {
      return false;
    }

    const fragment = convertFromRaw(rawContent).getBlockMap();

    const content = Modifier.replaceWithFragment(
      editorState.getCurrentContent(),
      editorState.getSelection(),
      fragment,
    );
    return EditorState.push(editorState, content, "insert-fragment");
  }

  return false;
};
