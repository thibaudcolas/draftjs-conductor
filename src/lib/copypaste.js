// @flow
import type { ElementRef } from "react";
import {
  Editor,
  EditorState,
  Modifier,
  convertToRaw,
  convertFromRaw,
  ContentState,
} from "draft-js";
import getFragmentFromSelection from "draft-js/lib/getFragmentFromSelection";

const copyHandlersRegistry = {};

export const registerCopySource = (ref: ElementRef<Editor>) => {
  const editorContainer = ref.editorContainer.parentNode;
  const key = ref.getEditorKey();

  const onCopy = (e) => {
    // Get clipboard content like Draft.js would.
    // https://github.com/facebook/draft-js/blob/37989027063ccc8279bfdc99a813b857549512a6/src/component/handlers/edit/editOnCopy.js#L34
    const fragment = getFragmentFromSelection(ref._latestEditorState);

    if (fragment) {
      const content = ContentState.createFromBlockArray(fragment.toArray());
      const serialisedContent = JSON.stringify(convertToRaw(content));

      e.target.setAttribute(
        "data-draftjs-conductor-fragment",
        serialisedContent,
      );

      // TODO Is it necessary to clean this up from the DOM?
      // TODO Look at React docs for HTML modifications to React-controlled elements.
      // TODO Find a way to remove the attribute as soon as the browser handled the copy event.
      window.setTimeout(() => {
        e.target.removeAttribute("data-draftjs-conductor-fragment");
      }, 100);
    }
  };

  editorContainer.addEventListener("copy", onCopy);
  copyHandlersRegistry[key] = onCopy;
};

export const unregisterCopySource = (ref: ElementRef<Editor>) => {
  const editorContainer = ref.editorContainer.parentNode;
  const key = ref.getEditorKey();
  const onCopy = copyHandlersRegistry[key];

  editorContainer.removeEventListener("copy", onCopy);
  delete copyHandlersRegistry[key];
};

export const handleDraftEditorPastedText = (
  ref: ElementRef<Editor>,
  text: string,
  html: ?string,
  editorState: EditorState,
) => {
  // Plain-text pastes are better handled by Draft.js.
  if (!html) {
    return false;
  }

  const parser = new DOMParser();
  const doc = parser.parseFromString(html, "text/html");
  const editor = doc.querySelector("[data-editor]");
  const sourceEditorKey = editor ? editor.getAttribute("data-editor") : "";
  const targetEditorKey = ref.getEditorKey();
  const serialisedContent = doc.querySelector(
    "[data-draftjs-conductor-fragment]",
  );

  if (
    sourceEditorKey &&
    sourceEditorKey !== targetEditorKey &&
    serialisedContent
  ) {
    const rawContent = JSON.parse(
      serialisedContent.getAttribute("data-draftjs-conductor-fragment") || "",
    );
    const clipboard = convertFromRaw(rawContent).getBlockMap();

    const content = Modifier.replaceWithFragment(
      editorState.getCurrentContent(),
      editorState.getSelection(),
      clipboard,
    );
    return EditorState.push(editorState, content, "insert-fragment");
  }

  return false;
};
