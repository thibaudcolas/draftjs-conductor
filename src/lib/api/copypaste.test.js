import {
  EditorState,
  convertFromRaw,
  convertToRaw,
  ContentState,
} from "draft-js";
import { registerCopySource, handleDraftEditorPastedText } from "./copypaste";

jest.mock("draft-js/lib/generateRandomKey", () => () => "a");
jest.mock("draft-js/lib/getFragmentFromSelection", () => () => ({
  toArray() {},
}));

describe("copypaste", () => {
  describe("registerCopySource", () => {
    it("registers and unregisters works for copy", () => {
      const editor = document.createElement("div");

      const copySource = registerCopySource({
        editor,
        _latestEditorState: EditorState.createEmpty(),
      });

      window.getSelection = jest.fn();
      editor.dispatchEvent(new Event("copy"));
      expect(window.getSelection).toHaveBeenCalled();

      copySource.unregister();

      window.getSelection = jest.fn();
      editor.dispatchEvent(new Event("cut"));
      expect(window.getSelection).not.toHaveBeenCalled();
    });

    it("works for cut", () => {
      const editor = document.createElement("div");

      const copySource = registerCopySource({
        editor,
        _latestEditorState: EditorState.createEmpty(),
      });

      window.getSelection = jest.fn();
      editor.dispatchEvent(new Event("cut"));
      expect(window.getSelection).toHaveBeenCalled();

      copySource.unregister();

      window.getSelection = jest.fn();
      editor.dispatchEvent(new Event("cut"));
      expect(window.getSelection).not.toHaveBeenCalled();
    });
  });

  /**
   * jsdom does not implement the DOM selection API, we have to do a lot of overriding.
   */
  describe("copy/cut listener", () => {
    it("no selection", () => {
      const editor = document.createElement("div");

      registerCopySource({
        editor,
        _latestEditorState: EditorState.createEmpty(),
      });

      window.getSelection = jest.fn(() => {
        return {
          rangeCount: 0,
        };
      });

      const event = new Event("copy");
      event.clipboardData = {};
      event.preventDefault = jest.fn();
      editor.dispatchEvent(event);

      expect(event.preventDefault).not.toHaveBeenCalled();
    });

    it("no clipboardData, IE11", () => {
      const editor = document.createElement("div");

      registerCopySource({
        editor,
        _latestEditorState: EditorState.createEmpty(),
      });

      window.getSelection = jest.fn(() => {
        return {
          rangeCount: 1,
        };
      });

      const event = new Event("copy");
      event.preventDefault = jest.fn();
      editor.dispatchEvent(event);

      expect(event.preventDefault).not.toHaveBeenCalled();
    });

    it("works", (done) => {
      const editor = document.createElement("div");

      registerCopySource({
        editor,
        _latestEditorState: EditorState.createEmpty(),
      });

      const content = {
        blocks: [
          {
            key: "a",
            type: "unstyled",
            text: "test",
          },
        ],
        entityMap: {},
      };

      jest
        .spyOn(ContentState, "createFromBlockArray")
        .mockImplementationOnce(() => {
          return convertFromRaw(content);
        });

      window.getSelection = jest.fn(() => {
        return {
          rangeCount: 1,
          toString: () => "toString selection",
          getRangeAt() {
            return {
              cloneContents() {
                return document.createElement("div");
              },
            };
          },
        };
      });

      const event = new Event("copy");
      event.preventDefault = jest.fn();
      event.clipboardData = {
        setData(type, data) {
          if (type === "text/plain") {
            expect(data).toBe("toString selection");
          } else if (type === "text/html") {
            expect(data).toMatchSnapshot();
            done();
          }
        },
      };
      editor.dispatchEvent(event);
    });
  });

  describe("handleDraftEditorPastedText", () => {
    it("no HTML", () => {
      const editorState = EditorState.createEmpty();
      expect(handleDraftEditorPastedText(null, editorState)).toBe(false);
    });

    it("HTML from other app", () => {
      const editorState = EditorState.createEmpty();
      const html = `<p>Hello, world!</p>`;
      expect(handleDraftEditorPastedText(html, editorState)).toBe(false);
    });

    it("HTML from draftjs-conductor", () => {
      const content = {
        blocks: [
          {
            data: {},
            depth: 0,
            entityRanges: [],
            inlineStyleRanges: [],
            key: "a",
            text: "hello,\nworld!",
            type: "unstyled",
          },
        ],
        entityMap: {},
      };
      let editorState = EditorState.createEmpty();
      const html = `<div data-draftjs-conductor-fragment='${JSON.stringify(
        content,
      )}'><p>Hello, world!</p></div>`;
      editorState = handleDraftEditorPastedText(html, editorState);
      expect(convertToRaw(editorState.getCurrentContent())).toEqual(content);
    });

    it("invalid JSON", () => {
      const editorState = EditorState.createEmpty();
      const html = `<div data-draftjs-conductor-fragment='{"blocks":[{"key"'><p>Hello, world!</p></div>`;
      expect(handleDraftEditorPastedText(html, editorState)).toBe(false);
    });
  });
});
