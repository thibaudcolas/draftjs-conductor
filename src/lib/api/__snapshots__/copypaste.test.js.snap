// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`copypaste copy/cut listener works 1`] = `"<div data-draftjs-conductor-fragment=\\"{&quot;blocks&quot;:[{&quot;key&quot;:&quot;a&quot;,&quot;text&quot;:&quot;test&quot;,&quot;type&quot;:&quot;unstyled&quot;,&quot;depth&quot;:0,&quot;inlineStyleRanges&quot;:[],&quot;entityRanges&quot;:[],&quot;data&quot;:{}}],&quot;entityMap&quot;:{}}\\" style=\\"white-space: pre-wrap;\\"><div></div></div>"`;
